src/skagent/model_analyzer.py

```diff
@@ -0,0 +1,274 @@
+        self.observables = set(observables or [])
+
+        # Flatten RBlock → list of DBlock(s)
+        self._blocks = []
+        self._walk_blocks()
+
+        # Storage
+        self.node_meta = {}
+        self._raw_deps = defaultdict(list)  # target → [sources…]
+        self._param_deps = defaultdict(set)  # param → {targets}
+        self._prev_deps = set()  # {(target, source), …}
+        self.edges = {"instant": [], "lag": [], "param": [], "shock": []}
+        self.formulas = {}
+        self.plates = {}
+
+    def _walk_blocks(self):
+        """Flatten RBlock into list of DBlock(s)."""
+        if isinstance(self.model, DBlock):
+            self._blocks = [self.model]
+        elif isinstance(self.model, RBlock):
+            for blk in self.model.blocks:
+                if isinstance(blk, DBlock):
+                    self._blocks.append(blk)
+                else:
+                    raise ValueError("Only one level of RBlock supported")
+        else:
+            raise ValueError("Model must be a DBlock or RBlock")
+
+    def analyze(self):
+        """Run the full analysis pipeline."""
+        self._collect_nodes()
+        self._collect_dependencies()
+
+        # Treat any referenced-but-unknown symbol as exogenous state
+        for tgt, deps in list(self._raw_deps.items()):
+            for src in deps:
+                if src not in self.node_meta:
+                    self.node_meta[src] = {
+                        "kind": "state",
+                        "agent": "global",
+                        "plate": None,
+                        "observed": False,
+                    }
+
+        self._identify_time_dependencies()
+        self._assemble_edges()
+        self._collect_formulas()
+        self._collect_plates()
+        return self
+
+    def _collect_nodes(self):
+        """Classify every variable and record its metadata."""
+        # 1) all LHS names in dynamics & reward
+        lhs = set()
+        for blk in self._blocks:
+            lhs |= set(blk.dynamics.keys())
+            lhs |= set(blk.reward.keys())
+
+        # 2) shocks
+        for blk in self._blocks:
+            for var, shock_def in blk.shocks.items():
+                agent = getattr(shock_def, "agent", "global") or "global"
+                self.node_meta[var] = {
+                    "kind": "shock",
+                    "agent": agent,
+                    "plate": agent if agent != "global" else None,
+                    "observed": False,
+                }
+
+        # 3) dynamics
+        for blk in self._blocks:
+            for var, rule in blk.dynamics.items():
+                kind = "control" if isinstance(rule, Control) else "state"
+                agent = getattr(rule, "agent", "global") or "global"
+                self.node_meta[var] = {
+                    "kind": kind,
+                    "agent": agent,
+                    "plate": agent if agent != "global" else None,
```

Comment:
I believe we've discussed that if the blocks are assigned to plates, that's different from variables being assigned to an agent.